util/auto_skip_compressor.cc

```diff
@@ -0,0 +1,118 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Creates mixed compressor wrapper which uses multiple compression algorithm
+// within same SST file.
+
+#include "util/auto_skip_compressor.h"
+
+#include <options/options_helper.h>
+
+#include "rocksdb/advanced_compression.h"
+#include "util/random.h"
+namespace ROCKSDB_NAMESPACE {
+
+int CompressionRejectionProbabilityPredictor::Predict() const {
+  return pred_rejection_percentage_;
+}
+
+bool CompressionRejectionProbabilityPredictor::Record(
+    Slice uncompressed_block_data, std::string* compressed_output,
+    const CompressionOptions& opts) {
+  if (compressed_output->size() >
+      (static_cast<uint64_t>(opts.max_compressed_bytes_per_kb) *
+       uncompressed_block_data.size()) >>
+      10) {
+    rejected_count_++;
+  } else {
+    compressed_count_++;
+  }
+  attempted_compression_count_++;
+  if (attempted_compression_count_ >= kWindowSize) {
+    pred_rejection_percentage_ = static_cast<int>(
+        rejected_count_ * 100 / (compressed_count_ + rejected_count_));
+    attempted_compression_count_ = 0;
+    compressed_count_ = 0;
+    rejected_count_ = 0;
+  }
+  return true;
+}
+AutoSkipCompressorWrapper::AutoSkipCompressorWrapper(
+    std::unique_ptr<Compressor> compressor, const CompressionOptions& opts,
+    const CompressionType type)
+    : CompressorWrapper::CompressorWrapper(std::move(compressor)),
+      opts_(opts),
+      type_(type) {}
+
+Status AutoSkipCompressorWrapper::CompressBlock(
+    Slice uncompressed_data, std::string* compressed_output,
+    CompressionType* out_compression_type, ManagedWorkingArea* wa) {
+  bool exploration =
+      Random::GetTLSInstance()->PercentTrue(kExplorationPercentage);
+  TEST_SYNC_POINT_CALLBACK(
+      "AutoSkipCompressorWrapper::CompressBlock::exploitOrExplore",
+      &exploration);
+  if (exploration) {
+    return CompressBlockAndRecord(uncompressed_data, compressed_output,
+                                  out_compression_type, wa);
+  } else {
+    auto predictor_ptr =
+        static_cast<AutoSkipCompressionContext*>(wa->get())->predictor_;
+    auto prediction = predictor_ptr->Predict();
+    if (prediction <= kProbabilityCutOff) {
+      // decide to compress
+      return CompressBlockAndRecord(uncompressed_data, compressed_output,
+                                    out_compression_type, wa);
+    } else {
+      // bypassed compression
```

Comment:
// bypass compression to be consistent with how you start with verb in "// decide to compress"